### ISSUE TITLE
[feature] apt-update for exist-db

### ISSUE THREAD

**Issue by eduarddrenth** [eXist-db/exist#4553]
For example when building an exist-db docker image it would be good if packages could be updated to the latest version.

See: https://exist-db.slack.com/archives/CG2MRUZ35/p1662555892413129

See: https://gist.github.com/line-o/8bab558790a532939432da5cd3ad8ef7

The gist can be a starting point, it should be reviewed and tested of course.

**Comment by line-o** [comment]
@eduarddrenth where or how would this be called?

**Comment by eduarddrenth** [comment]
Well, you could call it from your Dockerfile (bit complex example, sorry):

`RUN --mount=type=secret,id=commands,required,target=/exist/commands [ "java", "org.exist.start.Main", "client", "--no-gui",  "-l", "-u", "admin", "-P", "", "-F", "/exist/commands" ]`

**Comment by line-o** [comment]
I was able to remove the dependency from package-service completely in the updated gist.
Remaining ones are repo, http-client and semver. semver is the only library package, the others are built-in modules.
semver is bundled with eXist by default.

**Comment by eduarddrenth** [comment]
sounds good... and the hard coded package-repo?

**Comment by eduarddrenth** [comment]
OK, the file commandos holds arbitrary xquery, isn't part of exist but can
be supplied by any exist developer to be executed as part of a docker build
process. In this file I would like to call update-all

Op do 8 sep. 2022 17:02 schreef Juri Leino ***@***.***>:

**Comment by line-o** [comment]
Well, I will make it external but that likely won't help you much.
Would a request parameter help you?

**Comment by eduarddrenth** [comment]
> Well, I will make it external but that likely won't help you much. Would a request parameter help you?

You refer to the hard coded url I suppose? I don't know, can be in a global property/config? Otherwise perhaps a parameter (not a request parameter I would say) to the function...

**Comment by line-o** [comment]
> OK, the file commandos holds arbitrary xquery, isn't part of exist but can
be supplied by any exist developer to be executed as part of a docker build
process. In this file I would like to call update-all

@eduarddrenth is the commands file documented somewhere? When is this called? At which point in the lifetime of the container? At startup? _Before_ autodeploy is read?

Why is it not commands.xq - is that an API endpoint?

Assuming there is an XQuery main module "commands" that is evaluated each time the docker container starts.
1. If it is read before packages in autodeploy are installed this will fail as semver will not be available reliably. So at least the call to the function must be conditional.
1. If you want to call a function with packageUri as a parameter, then the gist code needs to be refactored into a library module.

Looking at the above you will likely be better off just using the functions that packageService already exposes.

**Comment by line-o** [comment]
see also https://github.com/eXist-db/xst/issues/33

**Comment by line-o** [comment]
OK, if you are sure that packageservice is installed, then

- gather installed packages using packages:get-installed()
- feed that list to packages:get-remote()
- filter the list (each item that has both available and installed)
- for each call packages install

**Comment by eduarddrenth** [comment]
> see also [eXist-db/xst#33](https://github.com/eXist-db/xst/issues/33)

good to know, but will not work within a (distroless) docker build of exist

**Comment by eduarddrenth** [comment]
> OK, if you are sure that packageservice is installed, then
> 
>     * gather installed packages using packages:get-installed()
> 
>     * feed that list to packages:get-remote()
> 
>     * filter the list (each item that has both available and installed)
> 
>     * for each call packages install

Isn't that what your update-all gist does? Or do you suggest a different approach here?

**Comment by line-o** [comment]
> > see also [eXist-db/xst#33](https://github.com/eXist-db/xst/issues/33)
> 
> good to know, but will not work within a (distroless) docker build of exist

With xst installed on the host machine you can "remote-control" an eXistdb instance running in docker.

**Comment by line-o** [comment]
> Isn't that what your update-all gist does? Or do you suggest a different approach here?

Not anymore as the gist does no longer depend on package service being installed.
But for you and for know you could just use the gist and modify it to your needs, right?
Or is it missing something?

**Comment by eduarddrenth** [comment]
> > > see also [eXist-db/xst#33](https://github.com/eXist-db/xst/issues/33)
> > 
> > 
> > good to know, but will not work within a (distroless) docker build of exist
> 
> With xst installed on the host machine you can "remote-control" an eXistdb instance running in docker.

Nice!

**Comment by eduarddrenth** [comment]
> > Isn't that what your update-all gist does? Or do you suggest a different approach here?
> 
> Not anymore as the gist does no longer depend on package service being installed. But for you and for know you could just use the gist and modify it to your needs, right? Or is it missing something?

Thanks, in due time I'll see how to integrate in my build/deploy pipelines

**Comment by eduarddrenth** [comment]
One more question, is the idea now to let "package update fucntionality" be a part of xst and not to add an update-all function to the package-service? This would somewhat force users to implement update functionality on docker hosts, which may be undesirable.

**Comment by adamretter** [comment]
> when building an exist-db docker image it would be good if packages could be updated to the latest version

When you build the Docker image of eXist-db it is always the case that the latest versions of both:
1. Linux packages for Debian are downloaded ands used
2. The latest version of Xar files, for the version of eXist-db that you are building, are downloaded and used

**Comment by line-o** [comment]
This feature is not implemented in exist-db (at runtime) yet.

**Comment by adamretter** [comment]
@line-o But I think the issue was about build time an not runtime. @eduarddrenth said this in the original enquiry:

> when building an exist-db docker image

**Comment by eduarddrenth** [comment]
> > when building an exist-db docker image it would be good if packages could be updated to the latest version
> 
> When you build the Docker image of eXist-db it is always the case that the latest versions of both:
> 
>     1. Linux packages for Debian are downloaded ands used
> 
>     2. The latest version of Xar files, for the version of eXist-db that you are building, are downloaded and used

Yes, I know, but I build my own docker apps based on an exist-db build from docker-hub (FROM exist-db...). In that situation updates aren't downloaded and that cannot be done in an easy way (exist-apt upgrade).

**Comment by adamretter** [comment]
@eduarddrenth If you use Maven to build your images from the official eXist-db images, it is fairly trivial to have it download the latest Xar files for you.

**Comment by line-o** [comment]
The feature to update all currently installed packages to their newest version within an instance of exist - at runtime - is a useful feature on its own regardless if @eduarddrenth's use-case can be achieved at build time or not. This is why I would like to keep this issue open.

**Comment by eduarddrenth** [comment]
Good to hear, I'll look into that, perhaps it's an option for me

Op zo 17 sep. 2023 13:49 schreef Adam Retter ***@***.***>:
